src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -24,12 +26,335 @@
+ * true on that case. Also, the deadlockPath is filled with the transaction
+ * nodes that form the cycle.
+ */
+static bool
+CheckDeadlockForTransactionNode(TransactionNode *startingTransactionNode,
+								TransactionNode **transactionNodeStack,
+								List **deadlockPath)
+{
+	List *waitingTransactionNodes = list_copy(startingTransactionNode->waitsFor);
+	int currentStackDepth = 0;
+	ListCell *waitingTransactionCell = NULL;
+
+	/*
+	 * We keep transactionNodeStack to keep track of the deadlock paths. At this point,
+	 * adjust the depth of the starting node and set the stack's first element with
+	 * the starting node.
+	 */
+	startingTransactionNode->currentStackDepth = currentStackDepth;
+	transactionNodeStack[currentStackDepth] = startingTransactionNode;
+
+	/* iterate on the children of the starting node and set their depths */
+	foreach(waitingTransactionCell, waitingTransactionNodes)
+	{
+		TransactionNode *waitingTransactionNode =
+			(TransactionNode *) lfirst(waitingTransactionCell);
+
+		waitingTransactionNode->currentStackDepth =
+			startingTransactionNode->currentStackDepth + 1;
+	}
+
+	/* traverse the graph and search for the deadlocks */
+	while (waitingTransactionNodes != NIL)
+	{
+		TransactionNode *waitingTransactionNode =
+			(TransactionNode *) linitial(waitingTransactionNodes);
+		ListCell *currentWaitForCell = NULL;
+
+		waitingTransactionNodes = list_delete_first(waitingTransactionNodes);
+
+		/* cycle found, let the caller know about the cycle */
+		if (waitingTransactionNode == startingTransactionNode)
+		{
+			FindDeadlockPath(waitingTransactionNode, transactionNodeStack, deadlockPath);
+
+			return true;
+		}
+
+		/* don't need to revisit the node again */
+		if (waitingTransactionNode->transactionVisited)
+		{
+			continue;
+		}
+
+		waitingTransactionNode->transactionVisited = true;
+
+		/* set the stack's corresponding element with the current node */
+		currentStackDepth = waitingTransactionNode->currentStackDepth;
+		transactionNodeStack[currentStackDepth] = waitingTransactionNode;
+
+		/* as we traverse outgoing edges, increment the depth */
+		currentStackDepth++;
+
+		/* prepend to the list to continue depth-first search */
+		foreach(currentWaitForCell, waitingTransactionNode->waitsFor)
+		{
+			TransactionNode *waitForTransaction =
+				(TransactionNode *) lfirst(currentWaitForCell);
+
+			/*
+			 * We add both the waiting node and its parent to simplify finding the
+			 * deadlock path.
+			 */
+			waitingTransactionNodes =
+				lcons(waitingTransactionNode, waitingTransactionNodes);
+
+			waitForTransaction->currentStackDepth = currentStackDepth;
+			waitingTransactionNodes = lcons(waitForTransaction, waitingTransactionNodes);
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * FindDeadlockPath gets the transaction node on which the cycle is found. Later,
+ * we add the depth number of elements of the cycled node from the transaction
+ * stack to the deadlockPath.
+ */
```

Comment:
The comment describes the manner, rather than the result. I think it'd be much more helpful if it said "* returns a list of transactions involved in a distributed deadlock (i.e. a cycle in the graph)" or such.